invalid memory address or nil pointer dereference"

**Describe the bug:**

```
$ k -n garden get po dependency-watchdog-probe-5699594f8b-z8cd4
NAME                                         READY   STATUS    RESTARTS   AGE
dependency-watchdog-probe-5699594f8b-z8cd4   1/1     Running   4          101m
```

```
$ k -n garden logs dependency-watchdog-probe-5699594f8b-z8cd4 -p

I0924 09:18:44.628382       1 scaler.go:72] Update event on cluster: shoot--it--tmghj-0da
E0924 09:18:44.715691       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 62 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x18bbae0, 0x2b49eb0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x18bbae0, 0x2b49eb0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/gardener/dependency-watchdog/pkg/scaler.shootHibernationStateChanged(0xc000972bc0, 0xc0000b8d00, 0xa)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/utils.go:88 +0x341
github.com/gardener/dependency-watchdog/pkg/scaler.NewController.func1(0x1aa80e0, 0xc000972bc0, 0x1aa80e0, 0xc0000b8d00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/scaler.go:79 +0xff
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00dcc9dd8, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003e6740)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00dcc9f40, 0xdf8475800, 0x0, 0x42da01, 0xc0005c6cc0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000367a00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0004c18e0, 0xc00035d290)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17046a1]

goroutine 62 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x18bbae0, 0x2b49eb0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/gardener/dependency-watchdog/pkg/scaler.shootHibernationStateChanged(0xc000972bc0, 0xc0000b8d00, 0xa)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/utils.go:88 +0x341
github.com/gardener/dependency-watchdog/pkg/scaler.NewController.func1(0x1aa80e0, 0xc000972bc0, 0x1aa80e0, 0xc0000b8d00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/pkg/scaler/scaler.go:79 +0xff
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00dcc9dd8, 0x0, 0x0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0003e6740)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00dcc9f40, 0xdf8475800, 0x0, 0x42da01, 0xc0005c6cc0)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000367a00)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0004c18e0, 0xc00035d290)
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/tmp/build/80754af9/git-gardener_dependency-watchdog-master_master/tmp/src/github.com/gardener/dependency-watchdog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```


**Anything else we need to know?**:

dependency-watchdog version: `v0.6.0`

--- a/pkg/scaler/utils.go
+++ b/pkg/scaler/utils.go
@@ -19,6 +19,8 @@ import (
 	"strings"
 
 	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
 	gardenerv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
 	"github.com/ghodss/yaml"
 	"k8s.io/klog"
@@ -79,17 +81,9 @@ func shootHibernationStateChanged(old, new *gardenerv1alpha1.Cluster) bool {
 		return false
 	}
 
-	if oldShoot.Status.IsHibernated != newShoot.Status.IsHibernated {
-		return true
-	}
+	return doCheckShootHibernationStateChanged(oldShoot, newShoot)
+}
 
-	if (oldShoot.Spec.Hibernation == nil && newShoot.Spec.Hibernation != nil) ||
-		(oldShoot.Spec.Hibernation != nil && newShoot.Spec.Hibernation == nil) ||
-		(oldShoot.Spec.Hibernation.Enabled == nil && newShoot.Spec.Hibernation.Enabled != nil) ||
-		(oldShoot.Spec.Hibernation.Enabled != nil && newShoot.Spec.Hibernation.Enabled == nil) ||
-		(oldShoot.Spec.Hibernation.Enabled != newShoot.Spec.Hibernation.Enabled) {
-		return true
-	}
-
-	return false
+func doCheckShootHibernationStateChanged(oldShoot, newShoot *gardencorev1beta1.Shoot) bool {
+	return oldShoot.Status.IsHibernated != newShoot.Status.IsHibernated || gardencorev1beta1helper.HibernationIsEnabled(oldShoot) != gardencorev1beta1helper.HibernationIsEnabled(newShoot)
 }

--- a/pkg/scaler/utils_test.go
+++ b/pkg/scaler/utils_test.go
@@ -1,0 +1,69 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scaler
+
+import (
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	yes = true
+	no  = false
+)
+
+var _ = DescribeTable("doCheckShootHibernationStateChanged", func(oldSpec *gardencorev1beta1.Hibernation, oldStatus bool, newSpec *gardencorev1beta1.Hibernation, newStatus, expectChanged bool) {
+	var (
+		old = &gardencorev1beta1.Shoot{
+			Spec: gardencorev1beta1.ShootSpec{
+				Hibernation: oldSpec,
+			},
+			Status: gardencorev1beta1.ShootStatus{
+				IsHibernated: oldStatus,
+			},
+		}
+		new = &gardencorev1beta1.Shoot{
+			Spec: gardencorev1beta1.ShootSpec{
+				Hibernation: newSpec,
+			},
+			Status: gardencorev1beta1.ShootStatus{
+				IsHibernated: newStatus,
+			},
+		}
+	)
+	Expect(doCheckShootHibernationStateChanged(old, new)).To(Equal(expectChanged))
+},
+	Entry("disabled, false, disabled, false", &gardencorev1beta1.Hibernation{Enabled: &no}, false, &gardencorev1beta1.Hibernation{Enabled: &no}, false, false),
+	Entry("disabled, false, enabled, false", &gardencorev1beta1.Hibernation{Enabled: &no}, false, &gardencorev1beta1.Hibernation{Enabled: &yes}, false, true),
+	Entry("disabled, true, disabled, true", &gardencorev1beta1.Hibernation{Enabled: &no}, true, &gardencorev1beta1.Hibernation{Enabled: &no}, true, false),
+	Entry("disabled, true, enabled, true", &gardencorev1beta1.Hibernation{Enabled: &no}, true, &gardencorev1beta1.Hibernation{Enabled: &yes}, true, true),
+	Entry("enabled, false, disabled, false", &gardencorev1beta1.Hibernation{Enabled: &yes}, false, &gardencorev1beta1.Hibernation{Enabled: &no}, false, true),
+	Entry("enabled, false, enabled, false", &gardencorev1beta1.Hibernation{Enabled: &yes}, false, &gardencorev1beta1.Hibernation{Enabled: &yes}, false, false),
+	Entry("enabled, true, disabled, true", &gardencorev1beta1.Hibernation{Enabled: &yes}, true, &gardencorev1beta1.Hibernation{Enabled: &no}, true, true),
+	Entry("enabled, true, enabled, true", &gardencorev1beta1.Hibernation{Enabled: &yes}, true, &gardencorev1beta1.Hibernation{Enabled: &yes}, true, false),
+	Entry("nil, false, nil, false", nil, false, nil, false, false),
+	Entry("nil, false, nil, true", nil, false, nil, true, true),
+	Entry("nil, false, non-nil, false", nil, false, &gardencorev1beta1.Hibernation{}, false, false),
+	Entry("nil, true, nil, false", nil, true, nil, false, true),
+	Entry("nil, true, nil, true", nil, true, nil, true, false),
+	Entry("nil, true, non-nil, true", nil, true, &gardencorev1beta1.Hibernation{}, true, false),
+	Entry("nil-enabled, false, non-nil-enabled, false", &gardencorev1beta1.Hibernation{}, false, &gardencorev1beta1.Hibernation{Enabled: &yes}, false, true),
+	Entry("nil-enabled, true, non-nil-enabled, true", &gardencorev1beta1.Hibernation{}, true, &gardencorev1beta1.Hibernation{Enabled: &yes}, true, true),
+	Entry("non-nil, false, nil, false", &gardencorev1beta1.Hibernation{}, false, nil, false, false),
+	Entry("non-nil, true, nil, true", &gardencorev1beta1.Hibernation{}, true, nil, true, false),
+	Entry("non-nil-enabled, false, nil-enabled, false", &gardencorev1beta1.Hibernation{Enabled: &yes}, false, &gardencorev1beta1.Hibernation{}, false, true),
+	Entry("non-nil-enabled, true, nil-enabled, true", &gardencorev1beta1.Hibernation{Enabled: &yes}, true, &gardencorev1beta1.Hibernation{}, true, true),
+)

